existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,318 @@
+
+## Universal quantification and `ExplicitForAll`
+
+By default, all type variables are considered universally quantified.
+This means that every time you use a function like `id`, you can _instantiate_ the universally quantified type variable `a` to (i.e., replace it with) any type.
+
+### Usage
+
+Let's try to add `forall` manually.
+The syntax is very simple – at the beginning of function or instance declaration, before any constraints or arguments are used, use
+the quantifier to introduce all type variables you want to use later:
+
+```Haskell
+func :: forall a b c m. a -> b -> m c -- OK, it compiles
+```
+Here we introduce four variables.
+All type variables that you introduce in your definition should belong to one `forall` exactly, though they may not belong to the same one. (I will explain it further.)
+
+```haskell
+func :: forall a b c.   a -> b -> m c -- Error, type variable `m` is not introduced
+```
+
+Let's add some constraints.
+For example, we might want to specify the `m` variable here as an instance of `Monad` typeclass.
+
+```Haskell
+func :: forall a b c m. Monad m => a -> b -> m c
+```
+You might want to ask, what's so special about this syntax?
+Nothing changed when we added the quantifier because it was already there (although implicit).
+However, making it explicit allows us to do some new things.
+
+### Examples
+
+#### Order of type variables
+
+First of all, you can now change the order of type variables.
+Let's imagine you want to define a function with 10 type variables (yes, there are real-life [examples](https://hackage.haskell.org/package/leancheck-0.9.10/docs/Test-LeanCheck-Utils-TypeBinding.html#v:-45--62--62--62--62--62--62--62--62--62--62--62--62-:) of such functions), and you know that when you use it, you want to specify the type of the last argument explicitly.
+
+Here and in the future examples, we will use the `TypeApplication` extension, which allows you to pass types as function arguments via the `@` sign. (If specifying concrete type is unnecessary, you can use the type wildcard `@_`.)
+
+By default, the order of type variables that you assign types to is the same as the order in which they are introduced in the function definition.
+
+Without `forall`, you would write something like this:
+
+```Haskell
+veryLongFunction :: a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @_ @_ @_ @_ @_ @_ @_ @_ @_ @Integer ...
+```
+Quite long, right?
+However, this can be simplified with explicit `forall`:
+
+```Haskell
+veryLongFunction :: forall j a b c d e f g h i. a -> b -> c -> d -> e -> f -> g -> h -> i -> j
+veryLongFunction = ...
+
+func = veryLongFunction @Integer ...
+```
+
+Since `j` is the first type variable in the declaration of `veryLongFunction`, you can specify only it.
+This is useful not only in such big examples but also when your function requires any of its arguments to be specified on usage.
```

Comment:
Is is obvious what "on usage" means to the reader?

Reply:
@PruStephan This is still unanswered.

Reply:
No it is not (At least I understood it only with third attempt). I thought I described it somewhere. Basically we are talking about on-usage (on-call) meaning for universal quantification and on-definition for existential types. On-usage means that when you call a function, you can specify type for it's arguments, like we did with TypeApplication. on-definition means that you can choose type for a existentially quantified variable when you define a function, and only there. Also, this concept looses meaning for existentially quantified datatypes

Reply:
Agreed, I also didn't understand what that meant.
I ended up deleting that sentence in https://github.com/serokell/blog-posts/pull/4/commits/27ef8ba24b79f17e56a214f51fe7606d434821ae because I believe the same idea was already explained a few paragraphs above, where I explained `TypeApplications` and instantiation in more detail.